src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -948,22 +985,54 @@ public SolutionCompilationState WithAnalyzerConfigDocumentText(
     }
 
     /// <inheritdoc cref="Solution.WithDocumentSyntaxRoots(ImmutableArray{ValueTuple{DocumentId, SyntaxNode}}, PreservationMode)"/>
+    /// <remarks>
+    /// This method supports updating ordinary documents and source generated documents.
+    /// </remarks>
     public SolutionCompilationState WithDocumentSyntaxRoots(ImmutableArray<(DocumentId documentId, SyntaxNode root)> syntaxRoots, PreservationMode mode)
     {
+        var (ordinaryDocuments, sourceGeneratedDocuments) = GetOrdinaryAndSourceGeneratedDocuments();
+
         return UpdateDocumentsInMultipleProjects<DocumentState, SyntaxNode, PreservationMode>(
-            syntaxRoots,
+            ordinaryDocuments,
+            sourceGeneratedDocuments,
             arg: mode,
             static (oldDocumentState, root, mode) =>
                 oldDocumentState.TryGetSyntaxTree(out var oldTree) && oldTree.TryGetRoot(out var oldRoot) && oldRoot == root
-                ? oldDocumentState
-                : oldDocumentState.UpdateTree(root, mode));
+                    ? oldDocumentState
+                    : oldDocumentState.UpdateTree(root, mode));
+
+        (ImmutableArray<(DocumentId, SyntaxNode)>,
+         ImmutableArray<(SourceGeneratedDocumentIdentity, DateTime, SourceText, SyntaxNode?)>) GetOrdinaryAndSourceGeneratedDocuments()
+        {
+            if (!syntaxRoots.Any(static t => t.documentId.IsSourceGenerated))
+                return (syntaxRoots, []);
+
+            using var _1 = ArrayBuilder<(DocumentId, SyntaxNode)>.GetInstance(capacity: syntaxRoots.Length, out var ordinaryDocuments);
+            using var _2 = ArrayBuilder<(SourceGeneratedDocumentIdentity, DateTime, SourceText, SyntaxNode?)>.GetInstance(out var sourceGeneratedDocuments);
+            foreach (var doc in syntaxRoots)
+            {
+                if (!doc.documentId.IsSourceGenerated)
+                {
+                    ordinaryDocuments.Add(doc);
+                }
+                else if (TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(doc.documentId) is { Identity: var identity })
+                {
+                    // Source generated documents always do a full parse, and source generator authors are only allowed to provide
+                    // strings, so we follow suit here and just take the text from the root.
+                    sourceGeneratedDocuments.Add((identity, DateTime.UtcNow, doc.root.GetText(), doc.root));
```

Comment:
See other comment about where I wish both text and root were nullable and one would pass through: in this case then we'd defer this GetText() until the actual creation of the state, since I'd hope nobody is actually thinking this text is available at that point. And since this is calling GetText on SyntaxRoot rather than on the tree (which has a magic deferred implementation) we're realizing the text all at once.